logic/transaction.js

```diff
@@ -769,7 +769,7 @@ Transaction.prototype.undo = function (trs, block, sender, cb) {
 		__private.types[trs.type].undo.call(this, trs, block, sender, function (err) {
 			if (err) {
 				this.scope.account.merge(sender.address, {
-					balance: amount,
+					balance: -amount,
 					blockId: block.id,
 					round: modules.rounds.calc(block.height)
 				}, function (err) {
```

Comment:
If we use a different variable for error, we can pass back the reason for failure.
Like: 			
```
			if (err) {
				this.scope.account.merge(sender.address, {
					balance: -amount,
					blockId: block.id,
					round: modules.rounds.calc(block.height)
				}, function (err2) {
					return setImmediate(cb, err2 || err);
				});
```